### ISSUE TITLE
The "Releases" link redirects to BenchmarkNet

### ISSUE THREAD

**Issue by wobes1** [nxrighthere/ENet-CSharp#27]
https://i.imgur.com/5KoCCDq.png

Clicking on releases redirects to: https://github.com/nxrighthere/BenchmarkNet/releases
Insted of: https://github.com/nxrighthere/ENet-CSharp/releases

**Comment by wobes1** [comment]
Please delete the issue thread after fixing the link.

**Comment by wobes1** [comment]
Didn't you think to open a discord server? And there do 3 sub channels for each of your repos? 

Would be nice.

**Comment by nxrighthere** [comment]
Actually, I was thinking about it, and I'm thinking about something like a blog or Patreon with free articles/code and optionally paid rewards. But not sure that this is the direction I want to move. Not sure that my content would be interesting to read/watch since it's quite specific stuff, I think.

**Comment by wobes1** [comment]
Well, a blog would be nice. Of course if you want to do so. A discord server would help to develop a community that will be interested in your works. Discussing ideas, asking question. Up to you then =)

Anyways, keep going. You are doing amazing and inspiring job.

**Comment by nxrighthere** [comment]
Thanks, I'll think about it. 🖖

**Comment by wobes1** [comment]
Don't want to create a new issue, or write to e-mail, so I will just posit it here =)

is it possible to add Packet.Create(IntPtr data)?

Motivation: 
To avoid passing byte[], pass IntPtr instead as packet data.

So what I am asking, is to add if it's possible.

`public void Create(IntPtr data, PacketFlags flags) {
	
		}
`

Thank you.

**Comment by nxrighthere** [comment]
Sure, I'll add an overload for this function.

**Comment by wobes1** [comment]
Thank you so much! I am inspired by the work you have done. Sorry, my knowledge of C++ is terrible so I can't do nothing there to write an overload by myself.
But the goal is to build an API based on Unity's ECS on top of ENet.

I hope it will be efficient. Thank you again!

**Comment by nxrighthere** [comment]
No problem. Make this changes: [1](https://www.diffchecker.com/diff#right-2) and [2](https://www.diffchecker.com/diff#right-4) and tell me how it works for ya.

**Comment by wobes1** [comment]
I guess difs are empty, or I am missing something?

**Comment by nxrighthere** [comment]
Ya, fixed.

**Comment by wobes1** [comment]
I'll add it later on and will get back to you.

Thank you for the quick replies and such incredible effort.

**Comment by nxrighthere** [comment]
You welcome.

**Comment by wobes1** [comment]
> You welcome.

Hi there. Do you have a moment? I am trying to connect Distruptor with ENet. Could you help me?
Here is my code so far.

https://gist.github.com/wobes1/f5969595454366848b206cb47b85a08f

**Comment by wobes1** [comment]
Got it right I guess. But why do we need RingBuffer?

https://gist.github.com/wobes1/43d64e2ea285975b668d74241bd26a53

**Comment by wobes1** [comment]
I feel so dumb. We use RingBuffer to queue events right?

**Comment by nxrighthere** [comment]
> I feel so dumb. We use RingBuffer to queue events, Inbound / Outbound right?

Right, you need to define messages for that and use queue for them. If your messages are small and doesn't contain any reference types, you can use struct for this.

```csharp
internal enum EventType {
	OnInitialized,
	OnDeinitialized,
	OnConnected,
	OnDisconnected,
	OnReceived,
	OnError
}

internal struct EventMessage {
	EventType event { get; set; }
	IntPtr buffer  { get; set; }
	Peer peer { get; set; }
	...
}
```

> So the question that bothers me, how to properly setup Timeout.

Do you mean timeout for processing data from queue? You don't need it, make both threads to process as much items as they can.

```csharp
while (queue.TryDequeue(out eventMessage)) {
	...
```

**Comment by nxrighthere** [comment]
If by timeout you mean interval for the service then set it to any update rate you want. 15 milliseconds is optional.

**Comment by nxrighthere** [comment]
You can leave it as is, default values are pretty balanced. But if you want to change it then, as soon as a peer is connected, on the server side set:

```csharp
	var timeout = 5000;

	peer.Timeout(ENet.throttleScale, timeout, timeout * 2); // 5 seconds minimum, 10 seconds maximum
```

**Comment by wobes1** [comment]
Thank you so much.

**Comment by wobes1** [comment]
Hello. You have suggested to create while loop based on event queue so the service method have to be inside the loop. But what if TryDequeue will return null, it means that Service won't be called? 
Or what is the proper way to setup consumer network thread? 

Cheers.

**Comment by nxrighthere** [comment]
Dequeue and the service work independently of each other. Firstly, you dequeue all inbound messages as much as possible, and then call the service to fill the outbound queue. Both inside a loop, the service will be used to control the update interval of the loop using the timeout parameter.

**Comment by wobes1** [comment]
Oh so the ENet Won't  block the while(deque) loop even with timeout set to 15? That what you meant by non blocking que?

**Comment by wobes1** [comment]
As much as possible? But won't it be 1 event per 15 millisecond if timeout is set to 15?

Is the snip right? 

while( try dequeue) 
{

// send a packet 
host.Service

}

**Comment by wobes1** [comment]
Oh. I think I understood. 
So 2 loops.

1. Main thread while dequeue()
2. Net thread while(true( { host.Service }

But I remember you mentioned, that when you call Send, you put in a queue first. Correct me if I'm wrong.

**Comment by nxrighthere** [comment]
Nope. Everything this stuff in the same thread, it should be:

```csharp
while (functionQueue.TryDequeue(out functionObject)) {
	...
}
```

then

```csharp
host.Service(1000 / tickRate, out netEvent);

switch (netEvent.Type) {
	...
```

or

```csharp
while (host.Service(0, out netEvent) > 0) {
	switch (netEvent.Type) {
		...

Thread.SpinWait(1);
```

**Comment by wobes1** [comment]
Sorry if I am annoying you. But I can't wrap my head around it.

Ok. So I am following your disrutptor queue approach and single producer  - single consumer.
Producer is a game thread. Consumer is a network thread.

I need 2 Queues. 1 st is From Game thread => network thread (to queue send event?)
2nd Network thread => game thread (to queue ENet received packets) and process them on the game thread.

The code in game loop then should be like

``while (outputQueeue.TryDequeue(out outputMessage))
 {
	// Process an ENet packet on game loop.
}

The network thread code:

```
new Thread(() =>
             {
                 while (inputMessage.TryDequeue(out inputMessage))
                 {
                    // send inputMessage

                     library.Host.Service(15, out ENet.Event netEvent);

                     switch (netEvent.Type)
                     {
                         //
                     }
                 }
             });
        }


```

**Comment by nxrighthere** [comment]
Separate the dequeue loop from the service.

**Comment by wobes1** [comment]
> Separate the dequeue loop from the service.

Gotcha. 
Thanks a lot. Much appreciate your patient.

**Comment by wobes1** [comment]
> Separate the dequeue loop from the service.

Won't the service block the consumer thread? Or disruptor works as unblockable?

**Comment by nxrighthere** [comment]
It's non-blocking queue based on memory barriers, and if you take a look into profiler, you will notice that there are no pauses from enqueuing/dequeuing messages.

**Comment by wobes1** [comment]
Hi there again. So far I got everything working correctly. Here is a snip of my network consumer thread. 
https://gist.github.com/wobes1/cccb3e1784ab84870599a223b0de7e15

So I can enqueue any event with 

`networkThreadQueue.Enqueue(new EventMessage { EventType = EventType.OnInitialized });`

is it right approach?
Thank you.

**Comment by nxrighthere** [comment]
Yea looks good so far, don't forget to add connect/disconnect messages as well. Now you can enqueue outbound messages for network events, and that's how you get stopless networking conveyor. 😺 

Also, as I mentioned in [this](https://github.com/nxrighthere/ENet-CSharp/issues/27#issuecomment-436885613) comment, you have two options on how to dispatch network events with the service. In the current approach that you are using now, ENet will dispatch one event per frame as fast as possible regardless of a timeout. Timeout here only needs to control blocking time and do not poll too long. Another way is to dispatch as many events as possible like that:

```csharp
while (host.Service(0, out netEvent) > 0) {
	switch (netEvent.Type) {
		...
	}
}

Thread.Sleep(1);
```

But this way the networking thread will consume more CPU resources, so you need to make some tests what's works best for you.

**Comment by wobes1** [comment]
You are my hero.
Thank you so much for your effort and your help with all of that.

As for the timeout. I will stick to 0 timeout approach for the client. But for the server, as I am planning to handle 500 connections I would stick to 15 milliseconds. Or I will try both 0 and will see the CPU consumption.

Your wrapper for ENet is an fantastic. 
Going to contribute another one on Bounty.

Cheers 🍻

**Comment by nxrighthere** [comment]
Thanks, I'm glad that you like it. 😄

**Comment by nxrighthere** [comment]
The beauty of non-blocking queues is that now you can parallelize your custom logic on top per channel basis. For example, you can create an array of queues for each channel, enqueue messages to an appropriate queue based on channel ID and process messages in multiple workers. There you can perform deserialization or any other computations and then push data to the game loop or somewhere else.

**Comment by wobes1** [comment]
> The beauty of non-blocking queues is that now you can parallelize your custom logic on top per channel basis. For example, you can create an array of queues for each channel, enqueue messages to an appropriate queue based on channel ID and process messages in multiple workers. There you can perform deserialization or any other computations and then push data to the game loop or somewhere else.

Hell yeah. My question is, what is the normal ammount of channels that has to be used? For such a survival game. That transmits position updates, some reliable events such as inventory drag & drop, and state updates of entities?

**Comment by nxrighthere** [comment]
Just follow this rule: let's say that you have multiple various actions like SpawnPlayer, DestroyPlayer, MovePlayer and so on. So you need to keep them in sequence because you can't move a player since he's not spawned and that message might come a bit earlier. So you are using one channel for this stuff, you want to preserve sequence for this actions. Each peer has 255 independently sequenced channels, so you separate logic in places where you don't need a sequence for their actions. Just always keep in mind: if you know that for a particular action disruption in sequence might break stuff, don't use an independent channel for that.

**Comment by wobes1** [comment]
Well, yes, thats obvious that I need sequence and reliability for such actions as Spawn, Destroy. For Move I might use unreliable flag but with server-side check if the player spawned | alive.

So, is it convenient . To use different channel for each PacketFlag?
In my case there would be 2 channels 

1. Reliable ordered
2. Unreliable

Right?

**Comment by nxrighthere** [comment]
It's absolutely fine, you can do it this way for sure.

**Comment by wobes1** [comment]
There is a bit of confusion to me. Channel and PacketFlag is two different things right?

But it's not recommended to use all PacketFlag types through one channel. Because lets say reliable would block unreliable sending?

**Comment by nxrighthere** [comment]
Nah, it will block only if a message is sequenced. If you are using the unsequenced flag, then the message will not be stalled in the channel.

**Comment by wobes1** [comment]
> Nah, it will block only if a message is sequenced. If you are using the unsequenced flag, then the message will not be stalled in the channel.

That what I am saying. If I am sending sequenced packet through one channel and then unreliable. Unreliable will be sent only after sequenced would be acknowledged. So thats' why I need 2 different channels for both of these types.

**Comment by wobes1** [comment]
> Nah, it will block only if a message is sequenced. If you are using the unsequenced flag, then the message will not be stalled in the channel.

Anyways, it's better to just have multiple channels dedicated for one PacketFlag :shipit:

**Comment by nxrighthere** [comment]
Unreliable will be send/received independently of acknowledged the reliable packet or not, so you can freely use one channel for both packets actually.

**Comment by wobes1** [comment]
Found this on ENet official forum. It seems I am right

> Since ENet will stall delivery of reliable packets to ensure proper sequencing, and consequently any packets of higher sequence number whether reliable or unreliable, in the event the reliable packet's predecessors have not yet arrived, this can introduce latency into the delivery of other packets which may not need to be as strictly ordered with respect to the packet that stalled their delivery.
> 
> To combat this latency and reduce the ordering restrictions on packets, ENet provides multiple channels of communication over a given connection. Each channel is independently sequenced, and so the delivery status of a packet in one channel will not stall the delivery of other packets in another channel.

**Comment by nxrighthere** [comment]
Well, you got this. 😄

**Comment by wobes1** [comment]
Hello Stanislav.

Have 1 more question. What can I do with uint data. I see that I can include it when connecting to a server. Could you give me some tip how to use it?

**Comment by nxrighthere** [comment]
Sure. You can use it to transmit simple stuff like a custom event ID, version number, and so on to reject connection regarding some conditions. The same thing with disconnection, you can use this data to transmit port number to seamlessly connect the client to another server on a different port if the current is full or to connect the client to the server that related to a zone in a world an so on.

**Comment by wobes1** [comment]
Thanks a lot! How about authorization? Is it plssible to pass some byte[] dara with connection? If not, I'll just write it on top.

Much appreciate that.

**Comment by wobes1** [comment]
Thanks a lot! How about authorization? Is it possible to pass some byte[] data with connection event? If not, I'll just write it on top.

Much appreciate that.

**Comment by wobes1** [comment]
Thanks a lot! How about authorization? Is it possible to pass some byte[] data with connection event? If not, I'll just write it on top.

Much appreciate that.

**Comment by nxrighthere** [comment]
Nope, this stuff should be used only after establishing a connection.

**Comment by wobes1** [comment]
Got it. What is your approach to assign an unique id to a packet? 
 
My approach is, 2 first bytes are the ushort packed header, the rest is serialized payload. But there would be unnecessary cutting of 2 first bytes in order to get the payload.

What would you do? Is there any unternal way of doing it on ENet side?

**Comment by wobes1** [comment]
Got it. What is your approach to assign an unique id to a packet?

My approach is, 2 first bytes are the ushort packed header, the rest is serialized payload. But there would be unnecessary cutting of 2 first bytes in order to get the payload.

What would you do? Is there any internal way of doing it on ENet side?

**Comment by nxrighthere** [comment]
Yep, this is the most trivial way, zero-copy buffer shifting. I'm doing this in the same way. Don't bother, 1 byte or 2 bytes doesn't matter, the most bandwidth consumptions come from data of the game, you know this.

You don't need any internal functions for this:
```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private void Pack(ushort id, byte[] data) {
	data[0] = (byte)(id & 0x00FF);
	data[1] = (byte)((id & 0xFF00) >> 8);
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
private void Unpack(byte[] buffer, int length) {
	for (int i = 0; i < length; i++) {
		buffer[i] = buffer[i + header];
	}
}
```

**Comment by wobes1** [comment]
Thanks a lot. That's true. But my project would have more than 255 different packets. That's why I need ushort. I was using Buffer.Copy. I will stick to your approach. 

Much appreciate your help.

**Comment by wobes1** [comment]
Have a moment? Do you think further optimization can be applied?
```C#
        private byte[] SerializePacket(Opcode opcode, IMessage protobufMessage)
        {
            int protobufLength = protobufMessage.CalculateSize();

            byte[] byteData = new byte[protobufLength + 2];

            AddPacketHeader((short)opcode, byteData);
            AddPacketPayload(protobufMessage, byteData, protobufLength);

            return byteData;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void AddPacketHeader(short opcode, byte[] data)
        {
            data[0] = (byte)(opcode & 0x00FF);
            data[1] = (byte)((opcode & 0xFF00) >> 8);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void AddPacketPayload(IMessage data, byte[] packetData, int dataLength)
        {
            Array.Copy(data.ToByteArray(), 0, packetData, 2, dataLength);
        }
```

**Comment by nxrighthere** [comment]
> byte[] packetData = new byte[messageLength + 2];

Instead of allocating buffer here, use [System.Buffers.ArrayPool](https://www.nuget.org/packages/System.Buffers/).

Use `Buffer.BlockCopy` instead of `Array.Copy` or use unsafe `Buffer.MemoryCopy` it's way more [faster](https://github.com/usausa/Work-Net-Benchmark/tree/master/CopyBenchmark).

**Comment by wobes1** [comment]
Many thanks.

**Comment by wobes1** [comment]
Hello Stanislav.

I carefully read your ScionNet readme file. There is a lot of useful information, thank you for that.
Found something interesting about RingBuffer capacities. I am using your medium parallelism degree pattern. What is your recommendation or formula to calculate capacity based on peer limit and update time? So far I am thinking about peerLimit * (1000 / updateTime).

Thank you.

**Comment by wobes1** [comment]
Much appreciate that.
